File: alchemist/src/main/java/it/unibo/alchemist/Alchemist.java
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
/**
 * 
 */
package it.unibo.alchemist;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Function;

import javax.swing.JFrame;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;
import org.apache.ignite.startup.cmdline.CommandLineStartup;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ch.qos.logback.classic.Level;
import it.unibo.alchemist.AlchemistRunner.Builder;
import it.unibo.alchemist.boundary.projectview.ProjectGUI;
import it.unibo.alchemist.cli.CLIMaker;
import it.unibo.alchemist.loader.Loader;
import it.unibo.alchemist.loader.YamlLoader;

/**
 * Starts Alchemist.
 * 
 */
public final class Alchemist {

    private static final Logger L = LoggerFactory.getLogger(Alchemist.class);
    private static final Map<String, Level> LOGLEVELS;
    private static final String HEADLESS = "hl";
    private static final String VARIABLES = "var";
    private static final String BENCHMARK = "bmk";
    private static final char BATCH = 'b';
    private static final char EXPORT = 'e';
    private static final char DISTRIBUTED = 'd';
    private static final char GRAPHICS = 'g';
    private static final char HELP = 'h';
    private static final char INTERVAL = 'i';
    private static final char NODE = 's';
    private static final char PARALLELISM = 'p';
    private static final char TIME = 't';
    private static final char YAML = 'y';

    static {
        final Map<String, Level> levels = new LinkedHashMap<>();
        levels.put("v", Level.INFO);
        levels.put("vv", Level.DEBUG);
        levels.put("vvv", Level.ALL);
        levels.put("q", Level.ERROR);
        levels.put("qq", Level.OFF);
        LOGLEVELS = Collections.unmodifiableMap(levels);
    }

    private Alchemist() {
    }

    /**
     * @param args
     *            the argument for the program
     * @param <T>
     *            concentration type
     */
    public static <T> void main(final String[] args) {
        final Options opts = CLIMaker.getOptions();
        final CommandLineParser parser = new DefaultParser();
        try {
            final CommandLine cmd = parser.parse(opts, args);
            setVerbosity(cmd);
            if (cmd.hasOption(NODE)) {
                CommandLineStartup.main(new String[] {cmd.getOptionValue(NODE)});
            }
            if (cmd.hasOption(HELP)) {
                final HelpFormatter formatter = new HelpFormatter();
                formatter.printHelp("java -jar alchemist-redist-{version}.jar", opts);
                System.exit(0);
            }
            Optional<Loader> loader = Optional.empty();
            if (cmd.hasOption(YAML)) {
                try (InputStream is = new FileInputStream(new File(cmd.getOptionValue(YAML)))) {
                    loader = Optional.of(new YamlLoader(is));
                } catch (IOException e) {
                    L.error("Unable to load the requested file.", e);
                }
            }
            if (loader.isPresent()) {
                final Builder<?, ?> simBuilder = new Builder<>(loader.get())
                        .setHeadless(cmd.hasOption(HEADLESS))
                        .setGUICloseOperation(JFrame.EXIT_ON_CLOSE);
                ifPresent(cmd, EXPORT, simBuilder::setOutputFile);
                ifPresent(cmd, GRAPHICS, simBuilder::setEffects);
                try {
                    ifPresent(cmd, INTERVAL, Double::parseDouble, simBuilder::setInterval);
                    ifPresent(cmd, TIME, Double::parseDouble, simBuilder::setEndTime);
                    final String[] varsUnderRun = cmd.getOptionValues(VARIABLES);
                    if (cmd.hasOption(BATCH)) {
                        if (cmd.hasOption(PARALLELISM)) {
                            try {
                                final int threads = Integer.parseUnsignedInt(cmd.getOptionValue(PARALLELISM));
                                simBuilder.setParallelism(threads);
                                L.info("Using " + threads + " thread(s).");
                            } catch (final NumberFormatException e) {
                                simBuilder.setParallelism(Runtime.getRuntime().availableProcessors());
                                L.warn("Invalid option for PARALLELISM parameter, back to default.");
                            }
                        }
                        if (cmd.hasOption(BENCHMARK)) {
                            simBuilder.setBenchmarkOutputFile(cmd.getOptionValue(BENCHMARK));
                        }
                        if (varsUnderRun == null) {
                            L.error("You must specify which variables you want the batch to run on.");
                            System.exit(1);
                        }
                        final String[] vars = Optional.ofNullable(cmd.getOptionValues(VARIABLES)).orElse(new String[0]);
                        if (vars.length == 0) {
                            L.info("Alchemist is in batch mode, but no variable is available.");
                            System.exit(2);
                        }

                        if (cmd.hasOption(DISTRIBUTED)) {
                            ifPresent(cmd, DISTRIBUTED, simBuilder::setRemoteConfig);
                        }
                        simBuilder.build().launch(vars);
                    } else {
                        simBuilder.build().launch();
                    }
                } catch (NumberFormatException e) {
                    L.error("A number was expected. " + e.getMessage());
                    System.exit(1);
                }
            } else {
                ProjectGUI.main();
            }
        } catch (ParseException e) {
            L.error("Your command sequence could not be parsed.", e);
        }
    }

    private static void ifPresent(final CommandLine cmd, final char opt, final Consumer<String> op) {
        ifPresent(cmd, opt, Function.identity(), op);
    }

    private static <R> void ifPresent(final CommandLine cmd, final char opt, final Function<String, R> fun, final Consumer<R> op) {
        getOpt(cmd, opt, fun).ifPresent(op);
    }

    private static <R> Optional<R> getOpt(final CommandLine cmd, final char opt, final Function<String, R> fun) {
        return Optional.ofNullable(cmd.hasOption(opt) ? fun.apply(cmd.getOptionValue(opt)) : null);
    }

    private static void setVerbosity(final CommandLine cmd) {
        boolean oneFound = false;
        for (final Entry<String, Level> entry: LOGLEVELS.entrySet()) {
            if (cmd.hasOption(entry.getKey())) {
                if (oneFound) {
                    /*
                     * If there are conflicting logging level requests, throw errors.
                     */
                    throw new IllegalArgumentException("Conflicting verbosity specification. Only one of " + LOGLEVELS.keySet() + " can be specified.");
                }
                oneFound = true;
                setLogbackLoggingLevel(entry.getValue());
            }
        }
    }

    private static void setLogbackLoggingLevel(final Level level) {
        final ch.qos.logback.classic.Logger root =
                (ch.qos.logback.classic.Logger) LoggerFactory.getLogger(ch.qos.logback.classic.Logger.ROOT_LOGGER_NAME);
        root.setLevel(level);
    }

}

File: alchemist/src/main/kotlin/it/unibo/alchemist/Alchemist.kt
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist

import ch.qos.logback.classic.Level
import ch.qos.logback.classic.Logger
import it.unibo.alchemist.cli.CLIMaker
import it.unibo.alchemist.launch.Launcher
import it.unibo.alchemist.launch.Validation
import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.CommandLineParser
import org.apache.commons.cli.DefaultParser
import org.apache.commons.cli.HelpFormatter
import org.apache.commons.cli.ParseException
import org.slf4j.LoggerFactory
import org.slf4j.helpers.NOPLoggerFactory
import java.lang.RuntimeException

/**
 * Starts Alchemist.
 */
object Alchemist {
    private const val HEADLESS = "hl"
    private const val VARIABLES = "var"
    private const val BATCH = 'b'
    private const val EXPORT = 'e'
    private const val DISTRIBUTED = 'd'
    private const val GRAPHICS = 'g'
    private const val HELP = 'h'
    private const val INTERVAL = 'i'
    private const val SERVER = 's'
    private const val PARALLELISM = 'p'
    private const val TIME = 't'
    private const val YAML = 'y'
    private val L = LoggerFactory.getLogger(Alchemist::class.java)
    private val launchers: List<Launcher> = ClassPathScanner
        .subTypesOf(Launcher::class.java, inPackage = "it.unibo.alchemist")
        .map { clazz ->
            val zeroAryConstructors = clazz.constructors
                .filter { it.parameterCount == 0 && it.canAccess(null) }
            if (zeroAryConstructors.size == 1) {
                zeroAryConstructors.first().newInstance() as Launcher
            } else {
                val instances = clazz.fields.filter {
                    it.name == "INSTANCE" &&
                    it.canAccess(null) &&
                    Launcher::class.java.isAssignableFrom(it.type)
                }
                if (instances.size == 1) {
                    instances.first().get(null) as Launcher
                } else {
                    throw IllegalStateException("Cannot instance or access an instance of $clazz")
                }
            }
        }
    private val logLevels = mapOf(
        "v" to Level.INFO,
        "vv" to Level.DEBUG,
        "vvv" to Level.ALL,
        "q" to Level.ERROR,
        "qq" to Level.OFF
    )
    /**
     * Set this to false for testing purposes.
     */
    private var isNormalExecution = true

    private inline fun <reified T : Number> CommandLine.hasNumeric(name: Char, converter: String.() -> T?): T? =
        getOptionValue(name)?.let {
            val value = converter(it)
            when {
                value == null ->
                    exitBecause("Not a valid ${T::class.simpleName}: $it", ExitStatus.NUMBER_FORMAT_ERROR)
                else -> value
            }
        }

    /**
     * @param args
     * the argument for the program
     */
    @JvmStatic
    fun main(args: Array<String>) {
        if (LoggerFactory.getILoggerFactory().javaClass == NOPLoggerFactory::class.java) {
            println("Alchemist could not load the output module (broken SLF4J depedencies?)") // NOPMD
            exitWith(ExitStatus.NO_LOGGER)
        }
        val opts = CLIMaker.getOptions()
        fun printHelp() = HelpFormatter().printHelp("java -jar alchemist-redist-{version}.jar", opts)
        val parser: CommandLineParser = DefaultParser()
        try {
            val cmd = parser.parse(opts, args)
            setVerbosity(cmd)
            val options = cmd.toAlchemist
            if (cmd.hasOption(HELP)) {
                printHelp()
                exitWith(ExitStatus.OK)
            }
            val (validLaunchers, invalidLaunchers) = launchers
                .map { it.validate(options) to it }
                .partition { (validation, _) -> validation is Validation.OK }
            when {
                validLaunchers.size == 1 -> validLaunchers.first().second(options)
                    .also { exitWith(ExitStatus.OK) }
                validLaunchers.size > 1 ->
                    L.error("Unable to select an execution strategy among ${validLaunchers.map {it.second} }")
                else -> {
                    L.error("No valid launchers.")
                    invalidLaunchers.forEach { (validation, launcher) ->
                        if (validation is Validation.Invalid) {
                            L.error("{}: {}", launcher::class.java.simpleName, validation.reason)
                        }
                    }
                }
            }
        } catch (e: ParseException) {
            L.error("Your command sequence could not be parsed.", e)
        }
        printHelp()
        exitWith(ExitStatus.INVALID_CLI)
    }

    /**
     * Call this method to enable testing mode, preventing Alchemist from shutting down the JVM.
     */
    fun enableTestMode() {
        isNormalExecution = false
    }

    private fun setVerbosity(cmd: CommandLine) {
        val verbosity = logLevels.filterKeys { cmd.hasOption(it) }.values
        when {
            verbosity.size > 1 -> exitBecause(
                    "Conflicting verbosity specification. Only one of ${logLevels.keys} can be specified.",
                    ExitStatus.MULTIPLE_VERBOSITY
                )
            verbosity.size == 1 -> setLogbackLoggingLevel(verbosity.first())
        }
    }

    private fun setLogbackLoggingLevel(level: Level) {
        val root =
            LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME) as Logger
        root.level = level
    }

    private fun exitWith(status: ExitStatus): Nothing {
        if (isNormalExecution) {
            System.exit(status.ordinal)
        }
        throw AlchemistWouldHaveExitedException(status.ordinal)
    }

    private fun exitBecause(reason: String, status: ExitStatus, exception: Exception? = null): Nothing {
        when {
            exception == null -> L.error(reason)
            else -> L.error(reason, exception)
        }
        exitWith(status)
    }

    private val CommandLine.toAlchemist: AlchemistExecutionOptions
        get() = AlchemistExecutionOptions(
            server = getOptionValue(SERVER),
            help = hasOption(HELP),
            batch = hasOption(BATCH),
            distributed = getOptionValue(DISTRIBUTED),
            endTime = hasNumeric(TIME, kotlin.String::toDoubleOrNull)
                ?: AlchemistExecutionOptions.defaultEndTime,
            export = getOptionValue(EXPORT),
            graphics = getOptionValue(GRAPHICS),
            headless = hasOption(HEADLESS),
            interval = hasNumeric(INTERVAL, kotlin.String::toDoubleOrNull)
                ?: AlchemistExecutionOptions.defaultInterval,
            parallelism = hasNumeric(PARALLELISM, kotlin.String::toIntOrNull)
                ?: AlchemistExecutionOptions.defaultParallelism,
            variables = getOptionValues(VARIABLES)?.toList()
                ?: emptyList(),
            configuration = getOptionValue(YAML)
        )

    private enum class ExitStatus {
        OK, INVALID_CLI, NO_LOGGER, NUMBER_FORMAT_ERROR, MULTIPLE_VERBOSITY
    }

    /**
     * This exception is thrown in place of calling [System.exit] when the simulator is used in debug mode.
     * The [exitStatus] returns the exit status the execution would have had.
     */
    data class AlchemistWouldHaveExitedException(
        val exitStatus: Int
    ) : RuntimeException("Alchemist would have exited with $exitStatus")
}
